src/compose/app.ts

```diff
@@ -631,11 +652,56 @@ export class App {
 		);
 	}
 
-	private generateContainerStep(current: Service, target: Service) {
+	// In the case where the Engine does not start the container despite the
+	// restart policy (this can happen in cases of Engine race conditions with
+	// host resources that are slower to be created but that a service relies on),
+	// we need to start the container after a delay.
+	private async requirementsMetForSpecialStart(
+		current: Service,
+		target: Service,
+	): Promise<boolean> {
+		// Shortcut the Engine inspect queries if status isn't exited
+		if (
+			current.status !== 'exited' ||
+			current.config.running !== false ||
+			target.config.running !== true
+		) {
+			return false;
+		}
+		let conditionsMetWithRestartPolicy = ['always', 'unless-stopped'].includes(
+			target.config.restart,
+		);
+
+		if (current.containerId != null && !conditionsMetWithRestartPolicy) {
+			const containerInspect = await inspectByDockerContainerId(
+				current.containerId,
+			);
+			// If the container has previously been started but exited unsuccessfully, it needs to be started.
+			// If the container has a restart policy of 'no' and has never been started, it
+			// should also be started, however, in that case, the status of the container will
+			// be 'Installed' and the state funnel will already be trying to start it until success,
+			// so we don't need to add any additional handling.
+			if (
+				target.config.restart === 'on-failure' &&
+				containerInspect.State.ExitCode !== 0
```

Comment:
To add to the point above, if we add `Status.ExitCode` to the `Service` class we can avoid using `async` in this method (and the call chain)